### PR TITLE
Fixed#30200 Added support to update array using indexes

### PR DESCRIPTION
**Pull request by nasirhjafri**
https://code.djangoproject.com/ticket/30200

- [x] Update `ArrayField` using indexes. i.e. `Foo.objects.update(arr__0=2)`
- [x] Update `JSONField` using indexes.
- [x] Update `HStoreField` using indexes
- [x] Update nested `ArraField` using indexes i.e. `Board.objects.update(pieces__0__0=4)`
for 
```python
class Board(models.Model):
    pieces = ArrayField(ArrayField(models.IntegerField()))
```

### CODE REVIEW

**Comment by nasirhjafri** [comment]
@felixxm Can you please review?

**Comment by auvipy** [comment]
you should uncheck the ticket patch needs improvemet

**Comment by nasirhjafri** [comment]
> you should uncheck the ticket patch needs improvemet

Looks like its already unchecked.

**Comment by auvipy** [comment]
i would suggest rebasing again on top of current master

File: django/db/models/sql/compiler.py [hunk 1]
```diff
@@ -1403,22 +1404,25 @@ def as_sql(self):
...
-                values.append('%s = %s' % (qn(name), placeholder % sql))
+                values.append(sql_placeholder % (placeholder % sql))
                 update_params.extend(params)
             elif val is not None:
-                values.append('%s = %s' % (qn(name), placeholder))
+                values.append(sql_placeholder % placeholder)
                 update_params.append(val)
             else:
-                values.append('%s = NULL' % qn(name))
+                values.append(sql_placeholder % 'NULL')
```
Comment (charettes) [inline django/db/models/sql/compiler.py]: It feels like the new or adjusted API should call a new `as_update_sql(lhs)` method on the lhs that passes the rhs and expect a `(str: sql, Tuple[Any]: params)` value to be returned. We have to make sure that by default this method raises by default so transforms like `update(date__year=2020)` fail but they could likely be implemented.

That would require `UpdateQuery.add_update_values` to solve the lookups and resolve both lhs and rhs while disallowing joins and thus aliasing. You can achieve that by sure making `UpdateQuery.alias_cols` is always `False`.

**Comment by charettes** [comment]
I agree with @felixxm that this doesn't seems to be heading in the right direction and that it should be based off the current transform API. I left a few comments about leads and ideas for enhanced tests.

File: django/contrib/postgres/fields/jsonb.py [hunk 2]
```diff
@@ -44,6 +44,13 @@ def __init__(self, verbose_name=None, name=None, encoder=None, **kwargs):
     def db_type(self, connection):
         return 'jsonb'
 
+    def get_sql_placeholder(self, quoted_name, key_lookups):
+        if not key_lookups:
+            return '{} = %s'.format(quoted_name)
+
+        keys = ','.join(key_lookups)
+        return "{name} = jsonb_set({name}, '{{{keys}}}', %s::jsonb)".format(name=quoted_name, keys=keys)
```
Comment (charettes) [inline django/contrib/postgres/fields/jsonb.py]: Another thing to consider when using `jsonb_set`; what should be done when trying to update a value a `NULL` value's key?

https://www.2ndquadrant.com/en/blog/jsonb_set_lax/

**Comment by felixxm** [comment]
Closing due to inactivity.

**Pull request closed without merging**